decoder.go

```diff
@@ -673,3 +695,85 @@ func indirect(v reflect.Value, decodingNull bool) (BinaryUnmarshaler, reflect.Va
 	}
 	return nil, v
 }
+
+func reflect_readArrayOfBytes(append bool, d *Decoder, l int, rv reflect.Value) error {
+	for i := 0; i < l; i++ {
+		n, err := d.ReadByte()
```

Comment:
We can save a lot of bounds checking here by
- doing a bounds check for the entire array
- allocating the entire slice
- then filling it

Same for all the other primitive arrays and slices

`reflect.Append` is worse than pre-allocation when the slice size is known because the end result is the same (in terms of memory usage), but the former is less efficient